### PR TITLE
v1.7.0 feat(sl*): add -c/--count flag + fix JSON summary counts (#245/#247)

### LINKED ISSUE #245: feat: add -c/--count flag to sl* listing commands to show file counts only
## Summary

Add a `-c/--count` flag to the sl* family listing commands (`sl`, `sls`, `slu`, `sla`, `slk`, `sli` — and `slc` when elifarley/hug-scm#244 lands) that prints only the number of matching files: the `grep -c` of file listings.

## Motivation

- Scripting today needs a pipe: `hug slc -q | wc -l` (or `hug slu | grep -c ...`). A native `--count` removes the pipe and the `-q` coupling, and mirrors the universal `grep -c` idiom.
- Precedent: `hug s` already has per-state count flags (`--staged`, `--unstaged`, `--untracked`, `--ignored` print counts to stdout) — the sl* family lacks the analog.
- Counts are the cheapest scriptable signal for "is there anything to act on?" (e.g. `hug slc -c` in a conflict-resolution hook: 0 → clean proceed, >0 → act).

## Proposed semantics

- `hug slu -c` → prints a single integer (newline-terminated) to stdout; nothing else on stdout (chatter stays on stderr, per the stdout/stderr discipline).
- Composes with pathspecs: `hug slc -c <path>` → count of conflicts under `<path>`.
- No matches → prints `0`, exit 0 (like `grep -c`).
- `-c` + `--json`: the unified JSON envelope already carries `summary.*` counts — decide whether `-c` is ignored, wins, or the combination errors (precedent: `hug wtl` errors on `--json --path-only`).
- `-c` + `-q` / trailing `hug s` summary: the count is the complete answer — likely `-c` suppresses the trailing summary; decide explicitly.

## Implementation sketch

- Each sl* script gains a `-c/--count` branch that pipes its list primitive to `wc -l` instead of the `list_files_with_status` rendering path (`hug slc -c` → `list_conflicted_files | wc -l`; `slu` → `list_unstaged_files`; …).
- Alternatively a `--count` mode inside the shared renderer — single source of truth, but touches the shared path used by all family commands.

## Out of scope (for now)

- Other listing families (`wtl`, `ll*`, `t*`) — generalize the flag later if the sl* precedent lands well.
- hg parity — the sl* family is git-only today.

## Related

- elifarley/hug-scm#244 — `hug slc` (in flight); this issue complements it.

### LINKED ISSUE #247: fix: unified JSON pipeline summary counts count comma-fragments, not files
## Summary

`summary.*` counts in the unified JSON pipeline (`output_json_status_unified` in `git-config/lib/hug-git-json`) count **comma-split fragments**, not files: one file inflates its count to 2.

## Reproduction

1 file with an unstaged modification → `hug slu --json` emits `"summary": {"unstaged": 2, "total": 2}` (and the `unstaged` array with the correct single object).

## Root cause

`collect_git_files_json` returns comma-joined JSON objects (e.g. `{"path": "f.txt", "status": "modified"}`), and `output_json_status_unified` splits them with `IFS=',' read -ra`, so each object becomes 2 array fragments (`{"path": "f.txt"` + ` "status": "modified"}`). The emitted ARRAY text is correct (re-joined with commas), but `summary.<type>` uses `${#arr[@]}` — the fragment count (2 per file; renames are 3 fields → 3).

## Affected commands

- `hug slu --json`, `hug sls --json`, `hug slk --json`, `hug sli --json` — inflated counts today.
- `hug s --json` is a **separate emitter** (`output_json_status_summary`, own counts via `wc -l`) — NOT affected.
- `hug slc --json` (elifarley/hug-scm#244) was given a **surgical object-count** for `summary.conflicted` so the new command is truthful; this issue is the legacy cleanup.

## Fix scope

Count objects, not fragments (e.g. count `"path"` markers per collection, or emit objects through a non-JSON-embedded delimiter). **This changes consumer-visible counts** for the four legacy commands — deliberate behavior change; update any docs and add a pinning test (none exists today — `tests/unit/test_json_edge_cases.bats` has no summary-count assertions).

### PR DESCRIPTION
**Pull request by elifarley**
## Summary

Two related changes to the `sl*` status-listing family, shipping together as **v1.7.0**:

**Feature — `-c/--count` flag (elifarley/hug-scm#245)**
Adds `-c/--count` to all seven `sl*` commands (`sl`, `sla`, `sls`, `slu`, `slk`, `sli`, `slc`): prints only the number of matching files — a single integer on stdout, `0` when none, exit `0` (the `grep -c` idiom). Removes the `| wc -l` pipe for scripting.

```bash
hug slc -c          # count conflicts (0 → clean, proceed)
hug slu -c src/     # count unstaged files under src/
if [[ $(hug slc -c) -gt 0 ]]; then ...; fi
```

**Bugfix — truthful JSON summary counts (elifarley/hug-scm#247)**
`slu`/`sls`/`slk`/`sli`/`slc`/`sl`/`sla --json` previously counted **comma-split JSON fragments** (a 2-field object inflated `summary.<type>` to 2), so one modified file reported `summary.unstaged = 2`. The collection contract now emits one JSON object per line and counts objects, so the summary matches the array length. This is a deliberate **consumer-visible behavior change**.

### Commits
- `feat(sl*)`: NUL-safe, Bash-4.0-safe `-c/--count` + JSON collection via line-per-object
- `test(sl*)`: pin mixed-filter JSON total + harden `-c` review findings
- `chore`: bump version to 1.7.0 + changelog
- `docs(meta)`: add `-c/--count` to `sl*` rows of hug-completion-reference

### Implementation notes
- New `count_files_with_status` engine (`git-config/lib/hug-select-files`) counts git's NUL-delimited records directly — so a filename containing a newline counts once (line-based counting would split it). It avoids `local -n` namerefs, so it runs on the **Bash 4.0–4.2 floor** the project supports (`docs/installation.md`), unlike `list_files_with_status` which routes through `convert_to_relative_paths`.
- Multi-state counts (`all`/`all+untracked`) use `git status --porcelain -z`, which emits ONE record per file (`MM` for a file both staged and unstaged) — deduplicated by construction.
- `collect_git_files_json` now prints one JSON object per line (was comma-joined); `mapfile` yields both the objects and the count. No nameref, no comma-split.

## Test Coverage

All new code paths have test coverage. Tests: 221 → 259 branch-relevant assertions (+36 across the new/updated suites).

- `tests/lib/test_hug-select-files.bats` — `count_files_with_status`: per-state counts, MM dedup, pathspec scoping, newline-filename NUL-safety, rename-counts-once, conflicted, unknown-state error branch (55 tests).
- `tests/unit/test_json_edge_cases.bats` — #247 regression pins for `slu`/`sls`/`slk`/`sli` single-filter counts, the rename-counts-as-1 pin, and a new `sl --json` **mixed-filter total** pin (the only multi-type path). Each pin was verified to FAIL on the pre-fix code (`2 2` / `2 2 4` vs the truthful `1 1` / `1 1 2`) (26 tests).
- `tests/unit/test_status_staging.bats` — `-c` behavior across all `sl*` commands, stdout cleanliness, mutual-exclusion error, pathspec no-match → `0` (162 tests).
- `tests/lib/test_hug_git_json.bats` — updated for the line-per-object collection contract (14 tests).

## Pre-Landing Review

**No CRITICAL findings.** Two specialists dispatched (Testing + Maintainability; Security/Performance skipped — pure Bash CLI, no backend/frontend/auth/migration/API scope). Full suite green in CI.

Auto-fixed (this PR):
- Corrected a factually-wrong test comment (claimed renames have "3 JSON fields"; they have 2 — `parse_file_to_json` emits only the new path).
- Corrected a wrong comment ("macOS wc emits trailing whitespace" → leading; now matches `git-s:289`).
- Added a porcelain-v1 version note to the record-skip check (prevents a future `--porcelain=v2` migration from silently breaking counts).
- Added the `sl --json` mixed-filter #247 pin and the `count_files_with_status` unknown-state error test.

Codex review (round 2, aca3b54) — two more defects found and fixed, both with regression tests verified to fail on the pre-fix code:
- (P1) `sla -c` / `sl -c` undercounted untracked files inside a new directory: `git status --porcelain -z` collapses them to `?? dir/`. Fix: pass `--untracked-files=all` when untracked files are requested.
- (P2) A staged rename whose old path is shaped like a status line (`M  old`) double-counted: the pos2 separator heuristic misread the bare old-path record. Fix: dropped the byte heuristic; the loop now tracks R/C records and consumes the following bare old-path record explicitly.

Considered and deferred (stable, low-value, or scope-expanding) — tracked as follow-up issues:
- DRY: the `-c` block is duplicated across the 6 `sl*` dispatchers. Factoring a `run_count_mode` helper is a real refactor; the duplication is clear and stable — not worth the churn here. Tracked in elifarley/hug-scm#258.
- `count_files_with_status` doesn't call `check_git_repo` (all callers do). Library invariant; defended in practice. Tracked in elifarley/hug-scm#259.

## Design Review

No frontend files changed — design review skipped.

## Eval Results

No prompt-related files changed — evals skipped.

## Scope Drift

Scope Check: CLEAN. Delivers exactly elifarley/hug-scm#245 (`-c/--count`) and elifarley/hug-scm#247 (JSON count fix). The two pre-existing `[Unreleased]` CHANGELOG entries (#234/#237, merged via an earlier PR but never promoted) correctly ship under 1.7.0 — they were unreleased.

## Plan Completion

No plan file detected — issue-driven (#245 + #247 are the spec). Both issues' acceptance criteria met:
- #245: all 7 `sl*` commands support `-c`; single integer on stdout; `0`/exit-0 on no match; pathspec composition; mutual-exclusion with `--json`; suppresses trailing summary.
- #247: all four legacy commands + `sl`/`sla` now count objects (files), with pinning tests.

## Verification Results

- All 259 branch-relevant BATS tests pass (json_edge_cases 26, status_staging 162, hug-select-files 57, hug_git_json 14).
- `make sanitize` clean (Bash format + lint, Python format + lint + typecheck).
- Smoke-tested end-to-end: `slu -c`, `sls -c`, `slk -c`, `sl -c` (dedup), `sla -c` (all+untracked), `sli -c` (0/exit-0), `-c --json` (errors), rename-counts-as-1, conflict-counts-as-1, `slu --json summary.unstaged = 1` (was 2).
- **One pre-existing local-only failure**: `test_commit.bats:239` ("surfaces commit hook failures") — fails only because this machine has a global `core.hooksPath` that overrides the repo's installed `commit-msg` hook (known issue class, tracked at elifarley/hug-scm#184). Green in CI (latest run on `main`: `success`). Unrelated to this branch (touches no commit code).

## Documentation

- `docs/meta/hug-completion-reference.md`: added `-c|--count` to the `sl` and `sls` rows of the Status (s*) section. The other five `sl*` rows (`sla`, `sli`, `slu`, `slk`, `slc`) cascade via "Args: as above" and pick up the addition transitively, matching the file's convention. (Primary docs — `docs/commands/status-staging.md` and `git-config/lib/README.md` — were already updated in the feature commit.)

Closes elifarley/hug-scm#245
Closes elifarley/hug-scm#247

## Test plan
- [x] All BATS tests pass (259 branch-relevant; full suite green save the documented local-only `test_commit.bats:239`)
- [x] `make sanitize` passes (format + lint + typecheck)
- [x] Manual smoke test of `-c` across all `sl*` commands + the #247 JSON fix

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: git-config/lib/hug-select-files [hunk 1]
```diff
@@ -6,6 +6,94 @@
...
+    # Select by porcelain code position; skip untracked (??) unless requested.
+    local want_index=false want_worktree=false want_untracked=false
+    case "$state" in
+    staged) want_index=true ;;
+    unstaged) want_worktree=true ;;
+    all) want_index=true; want_worktree=true ;;
+    all+untracked) want_index=true; want_worktree=true; want_untracked=true ;;
+    esac
+
+    count=$(git status --porcelain -z -- "${pathspecs[@]}" 2>/dev/null | while IFS= read -r -d '' rec; do
```
[the remark this answers is not in the document]
Comment (elifarley) [reply git-config/lib/hug-select-files]: Verified and fixed in aca3b54. Reproduced first: two untracked files under a new directory returned 1 from `hug sla -c` (the `?? newdir/` collapse) while `hug slk -c` correctly returned 2 — because the dedicated `untracked` state uses `git ls-files --others` (expands files) but the `all+untracked` state used `git status --porcelain -z` without `--untracked-files=all` (collapses the dir). Fix: the porcelain path now passes `--untracked-files=all` whenever `$want_untracked` is true. Added a regression test (`all+untracked expands untracked dir to its files`) that fails on the pre-fix code (1 vs 2).

File: git-config/lib/hug-select-files [hunk 2]
```diff
@@ -6,6 +6,94 @@
...
+      # XY prefix) — its pos2 is a filename char, not space/tab, so it is
+      # skipped here (the rename already counted once via its "R  new" record).
+      # IMPORTANT: this assumes porcelain v1 — `git status --porcelain -z`
+      # defaults to v1. Do NOT switch to `--porcelain=v2 -z`: v2's record layout
+      # ("XY<space>status...<space>path") moves the path separator off pos2 and
+      # would silently mis-count. `if` keeps this safe under `set -e`
+      # (a bare `&& continue` would abort).
+      if [[ "${rec:2:1}" != " " && "${rec:2:1}" != $'\t' ]]; then
+        continue
+      fi
```
[the remark this answers is not in the document]
Comment (elifarley) [reply git-config/lib/hug-select-files]: Verified and fixed in aca3b54. The pos2 separator heuristic was indeed the wrong invariant — it inferred record type from a filename byte, which is arbitrary. Reproduced: renaming a tracked file literally named "M  old" made `hug sls -c` return 2, because git emits "R  new\0M  old\0" and the bare old-path record's 3rd byte is a space, fooling the heuristic into reading it as a second staged record. Fix: dropped the pos2 heuristic entirely; the loop now tracks R/C records explicitly and consumes the following bare old-path record by name (`consume_rename_old` flag), robust to any old-path bytes. Added a regression test (`rename old-path mimicking a status record counts once`) that fails on the pre-fix code (2 vs 1).
